dropshot/src/handler.rs

```diff
@@ -99,7 +100,14 @@ pub struct RequestContext {
     pub log: Logger,
 }
 
-impl RequestContext {
+impl<Context: ServerContext> RequestContext<Context> {
+    /**
+     * Returns the server context state.
+     */
+    pub fn context(&self) -> Arc<Context> {
```

Comment:
Given this new mechanism of extracting the context, I actually wonder...

- Does it need to be wrapped in an Arc?
- When providing it back to a caller, do we need to provide **new** Arc object?

Seems like it should be possible to return either an `&Arc<Context>` or a `&Context` object directly.

Reply:
Cloning the Arc on each dispatch has a performance impact. `&Arc` would be the smaller change -- code would retain the ability to make its own clone of the `Arc` if required.

However, I'd bet that current handler functions are dropping the `Arc` at the end, which means `&Context` would be fine -- it has the bonus of being a simpler, clearer type and exposing fewer implementation details of the server. So, that's what I'd nudge toward.

Reply:
Agreed. I've replaced the usage here with `&Context` directly (in the callback, as well as in storage). Users can decide for themselves if they'd like to wrap this type in an `Arc`.

Reply:
I definitely like if we can drop the Arc here.  Does that work for API endpoint handler functions that are async and whose futures reference the context after they return?

Reply:
> I definitely like if we can drop the Arc here. Does that work for API endpoint handler functions that are async and whose futures reference the context after they return?

This is up to the caller. If they'd like this behavior, they can supply arguments such that `Context=Arc<MyObject>`, or any other smart-pointer wrapping of choice. However, if callers don't need this behavior, they can just supply `Context=MyObject`. The `context` function will be monomorphized to return whichever variant the caller requested.

Reply:
Makes sense.  Maybe we should have Context=Arc<Example> in the basic example, as I think that'll be the most common case by far.

Reply:
I updated the "example context" to use asynchronous functions, but it *still* didn't need an Arc because the endpoint functions are also async, so the lifetime wasn't a problem. I think the case where the Arc is necessary is where the server wants to act on data **completely independently** of endpoint functions; I made another example which increments a counter on a timer.

(I basically just want the example to have a "real" reason for using the Arc. Added in `module-shared-context.rs`)

Reply:
Sorry -- I was confused about the `Arc` around `RequestContext` vs. the user's context type.  This all makes more sense now!